hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/ECUnderReplicationCommandCreator.java

```diff
@@ -0,0 +1,239 @@
+   * @return Returns the key value pair of destination dn where the command gets
+   * executed and the command itself.
+   */
+  @Override
+  public Map<DatanodeDetails, SCMCommand> processAndCreateCommands(
+      final Set<ContainerReplica> replicas,
+      final ContainerReplicaPendingOps pendingOps,
+      final ContainerHealthResult result,
+      final int remainingMaintenanceRedundancy) {
+    ContainerInfo container = result.getContainerInfo();
+    ECReplicationConfig repConfig =
+        (ECReplicationConfig) container.getReplicationConfig();
+
+    final ECContainerReplicaCount replicaCount =
+        new ECContainerReplicaCount(container, replicas,
+            pendingOps.getPendingOps(container.containerID()),
+            remainingMaintenanceRedundancy);
+    ContainerHealthResult containerHealthResult =
+        new ECContainerHealthCheck().checkHealth(container, replicaCount);
+
+    LOG.debug("Handling under-replicated EC container: {}", container);
+
+    if (containerHealthResult.getHealthState()
+        != ContainerHealthResult.HealthState.UNDER_REPLICATED) {
+      LOG.info(
+          "The container {} with replicas {} is sufficiently replicated",
+          container.getContainerID(), replicaCount.getReplicas());
+      // Assert replicaCount also telling it's sufficiently replicated.
+      return null;
+    }
+    final ContainerID id = container.containerID();
+    try {
+     // State is UNDER_REPLICATED
+      List<ContainerReplicaOp> pendingOpsForContainer =
+          pendingOps.getPendingOps(id);
+      final List<DatanodeDetails> deletionInFlight = new ArrayList<>();
+      for (ContainerReplicaOp op : pendingOpsForContainer) {
+        if (op.getOpType() == ContainerReplicaOp.PendingOpType.DELETE) {
+          deletionInFlight.add(op.getTarget());
+        }
+      }
+      List<Integer> missingIndexes = replicaCount.unavailableIndexes(true);
+      // We got the missing indexes, this is excluded any decommissioning
+      // indexes. Find the good source nodes.
+      List<ContainerReplica> sources = replicas.stream().filter(r -> r
+          .getState() == StorageContainerDatanodeProtocolProtos
+          .ContainerReplicaProto.State.CLOSED)
+          // Exclude stale and dead nodes. This is particularly important for
+          // maintenance nodes, as the replicas will remain present in the
+          // container manager, even when they go dead.
+          .filter(r -> ReplicationManager
+              .getNodeStatus(r.getDatanodeDetails(), nodeManager).isHealthy())
+          .filter(r -> !deletionInFlight.contains(r.getDatanodeDetails()))
+          .collect(Collectors.toList());
+
+      if (missingIndexes.size() > 0 && !replicaCount.unRecoverable()) {
+        LOG.debug("Missing indexes detected for the container {}." +
+                " The missing indexes are {}", id, missingIndexes);
+        // We have source nodes.
+        if (sources.size() > 0) {
```

Comment:
Should the sources check be `>= ec_data_num` - as if we don't have dataNum sources, we cannot recover?

Reply:
Yeah, we should add enough nodes check here to void sending in sufficient sources.